File: ATListView/Classes/Proxy/DataSourceProxy.m
//
//  DataSourceProxy.m
//  Alamofire
//
//  Created by 凯文马 on 2018/3/2.
//

#import "DataSourceProxy.h"
#import <objc/runtime.h>

static NSMutableDictionary *dataSourceClassMap = nil;

@implementation DataSourceProxy

+ (instancetype)dataSourceWithListView:(id<UITableViewDataSource>)listView realDataSource:(id<UITableViewDataSource>)realDataSource forObject:(id)object
{
    NSString *clsN = NSStringFromClass([object class]);
    if (clsN == nil) {
        return [[DataSourceProxy alloc] initWithListView:listView realDataSource:realDataSource];
    }
    NSMutableDictionary *map = self.classMap;
    Class cls = map[clsN];
    if (cls) {
        return [[cls alloc] initWithListView:listView realDataSource:realDataSource];
    }
    
    cls = objc_allocateClassPair([DataSourceProxy class], [NSString stringWithFormat:@"%@_DelegateProxy",clsN].UTF8String, 0);
    map[clsN] = cls;
    return [[cls alloc] initWithListView:listView realDataSource:realDataSource];
}

+ (NSMutableDictionary *)classMap
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        dataSourceClassMap = [@{} mutableCopy];
    });
    return dataSourceClassMap;
}

- (instancetype)initWithListView:(id<UITableViewDataSource>)listView realDataSource:(id<UITableViewDataSource>)realDataSource
{
    self = [super init];
    if (self) {
        _listView = listView;
        _realDataSource = realDataSource;
    }
    return self;
}

- (BOOL)respondsToSelector:(SEL)aSelector
{
    if ([_realDataSource respondsToSelector:aSelector]) {
        return YES;
    } else if ([super respondsToSelector:aSelector]) {
        return YES;
    }
    return NO;
}


- (id)forwardingTargetForSelector:(SEL)aSelector
{
    return _realDataSource;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    return [_listView tableView:tableView cellForRowAtIndexPath:indexPath];
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [_listView tableView:tableView numberOfRowsInSection:section];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [_listView numberOfSectionsInTableView:tableView];
}

@end

File: ATListView/Classes/Proxy/DataSourceProxy.swift
//
//  DataSourceProxy.swift
//  ATListView
//
//  Created by 凯文马 on 21/12/2017.
//

import UIKit

class DataSourceProxy
: NSObject
, UITableViewDataSource
{
    
    weak var mainProxy : UITableViewDataSource?
    
    weak var secondProxy : UITableViewDataSource?
    
    func exchangeProxy() {
        if mainProxy == nil || secondProxy == nil {
            return
        }
        (mainProxy,secondProxy) = (secondProxy,mainProxy)
    }
    
    init(main: UITableViewDataSource?, second: UITableViewDataSource? = nil) {
        mainProxy = main
        secondProxy = second
    }
    
    override func responds(to aSelector: Selector!) -> Bool {
        return super.responds(to: aSelector) || (mainProxy?.responds(to: aSelector) ?? false)
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let proxy = secondProxy ?? mainProxy ?? self
        if !proxy.isEqual(self) {
            return proxy.tableView(tableView, numberOfRowsInSection: section)
        }
        return 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let proxy = secondProxy ?? mainProxy ?? self
        if !proxy.isEqual(self) {
            return proxy.tableView(tableView, cellForRowAt: indexPath)
        }
        return UITableViewCell()
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        var num = 1
        let proxy = secondProxy ?? mainProxy ?? self
        if !proxy.isEqual(self) {
             num = proxy.numberOfSections?(in: tableView) ?? 1
        }
        return num
    }
    
    override func forwardingTarget(for aSelector: Selector!) -> Any? {
        if mainProxy != nil && mainProxy!.responds(to: aSelector) {
            return mainProxy!
        } else if secondProxy != nil && secondProxy!.responds(to: aSelector) {
            return secondProxy!
        }
        return super.forwardingTarget(for: aSelector)
    }
    
    override var description: String {
        return debugDescription
    }
    
    override var debugDescription: String {
        return "UITableViewDataSource<\n\(String(describing: mainProxy)),\n\(String(describing: secondProxy))>"
    }
}

